### ISSUE TITLE
Better Dependency Injection

### ISSUE THREAD

**Issue by cable729** [graphql-dotnet/graphql-dotnet#97]
First of all, thank you for creating this awesome library. It's wonderful to be able ot use GraphQL in my .NET projects.

While working with this project, dependency injection has been a little bit of a sticking point. The two main issues I'm facing are:
1. I have to set my dependency injection function in the constructor. The limitation here is that I can't use a class function or variable as a parameter. I.E. I can't do something like this:

```
public MySchema(...) : base(ResolveDependency) { ... }

public GraphType ResolveDependency(Type type) { ... }
```
1. List types are currently being requested internally as `ListGraphType<T>`. Of course, I don't want to register every type of generic list I'd use. My workaround (the same concept as in the sample project) is as follows:

```
            : base(type =>
            {
                if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(ListGraphType<>))
                {
                    return (GraphType) Activator.CreateInstance(type);
                }

                return (GraphType) context.Resolve(type);
            })
```

On top of everything, doing dependency injection in this system just feels clunky. I wonder if we want to do something more like:

```
public interface IDependencyResolver
{
    public T Resolve<T>();
    public object Resolve(Type type);
}
```

Notice, instead of requiring to resolve to a `GraphType`, we resolve any type, and inject that into the constructor of the `GraphType`s we need.

**Comment by joemcbride** [comment]
Thanks for the feedback.

1. The `ctor` is just an overload.  The `Schema` class has a `ResolveType` property that the provided func gets assigned to.  You can re-assign that property on the class and pass it a class function.

2. The engine requests the type that the particular field needs, which can often be a non-null or a list.  Unfortunately, the current design requires all permutations of those types to be registered, which is kind of dumb.  I Agree with you here.

As per more generic dependency injection, let me talk a little about the purpose of the `ResolveType` func.

This is *framework* stuff.  The GraphQL engine is looking to get an instance of a type to work with.  In other GraphQL implementations, the types are new'd up right away vs. just registering the `Type`.  Since we're not doing that, we have to have some way to create the type, hence the `ResolveType` func.  In this *framework* sense, GraphQL only ever cares about `GraphType` instances - it won't ever care about anything else.

Part of the reason to have the resolver though is so you can provide dependencies in the `ctor` of your `GraphType` classes that will get resolved.  The engine always wants a `GraphType` in the end, though that `GraphType` may have other dependencies.

I have plans to move away from declaring and tracking `Type`'s and move to instances.  This can be seen in this spike branch https://github.com/graphql-dotnet/graphql-dotnet/pull/91 PR.  Doing so will eliminate the need for any type resolving from the framework and put that onus on the user.

One thing to keep in mind is that a GraphQL system only ever needs a single instance of a given `GraphType` in the system.  What happens right now is when the system "boots up" for the first time, it walks the dependency graph and creates a cached list of all of the types in the system.  It only does this once and then pulls the types from the cache.

Now that all that is said - do you have a scenario where you want some sort of dependency injection that is not constructor based on the `GraphType`?  Are you looking to do service-location in a field resolve method?

**Comment by tlil87** [comment]
Should be noted that this will also behave differently in the Conventions library, so will ping some details on that when the release is out. Just wanted to throw that in there.

**Comment by baluubas** [comment]
I'm struggling a bit with this too, that is, injecting dependencies to do the resolving.

What I currently do is some sort of poor man's dependency resolver thingy.

In the root type I set a static property on all types that require some sort of injections.

```
	public class RecruitQuery : ObjectGraphType
	{
		public RecruitQuery(IServiceProvider ioc)
		{
			Name = "Root";

			// Poor mans DI
			MeType._ioc = ioc;
			
			Field<MeType>(
				"me",
				"Root for queries based on the authenticated user",
				resolve: ctx => new MeObj()
			);
		}
	}
```

and in the MeType 

```
	public class MeType : ObjectGraphType
	{
		internal static IServiceProvider _ioc { get; set; }

		public MeType()
		{
			Name = "Me";

			Field<NonNullGraphType<UserType>>(
				"profile",
				"Fetches info for the autenticated user.",
				resolve: context => _ioc.GetService<GetAutenticatedUser>().Execute()
			);
		}
	}
}
```

It would be nice to if this would be build-in. Perhaps by registering a resolver (inheriting from some interface) when creating a field

```
	Field<NonNullGraphType<UserType>>("profile").ResolveWith<GetAutenticatedUser>();
```

**Comment by JarnoNijboer** [comment]
You could used ctor injection, so that the `IGetAuthenticatedUser` is injected into the ctor from `MeType`.  Then the resolve method would look like `resolve: ctx => getAuthenticatedUser.Execute()`
Such a solution doesn't force you to sprinkle your Ioc container throughout your GraphQL code.

An example of what this would look like:
```csharp
public class MeType : ObjectGraphType
{
    public MeType(IGetAuthenticatedUser getAutenticatedUser)
    {
        Name = "Me";

        Field<NonNullGraphType<UserType>>(
            "profile",
            "Fetches info for the autenticated user.",
            resolve: context => getAutenticatedUser.Execute()
        );
    }
}
```

**Comment by baluubas** [comment]
@JarnoNijboer  I'm quite sure that doesn't work. I might be out on limb here but I think the instance of `MeType` is created by the framework using `Activator.CreateInstance()` and the way it's used it requires a empty constructor. 

It can work for the root query but if you have nesting as I do you run in to this problem.

**Comment by JarnoNijboer** [comment]
@baluubas It will work, trust me 😉 Since you can pass a `Func<Type, GraphType>` into the ctor of the [Schema](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Types/Schema.cs#L46) class, as outlined by @joemcbride [above](https://github.com/graphql-dotnet/graphql-dotnet/issues/97#issuecomment-225059182)

The func you can pass in the ctor allows you to plugin your Ioc container, and to me the  default `Activator.CreateInstance` is just a fallback, for when you don't use a Ioc container at all.
An example of how I use StructureMap, in conjuction with GraphQL, it in one the projects I work with:
```csharp
using GraphQL.Types;
using StructureMap;

public class MySchema : Schema
{
    public MySchema(IContainer container) : this(type => (GraphType) container.GetInstance(type)) { }

    private MySchema(Func<Type, GraphType> resolveType)
        : base(resolveType)
    {
        ...
    }
}
```

**Comment by baluubas** [comment]
Ah, I totally read that wrong - missed that it could be used to resolve GraphType in that context. Yes, that does work beautifully - thanks for pointing me in the right direction @JarnoNijboer .

**Comment by cable729** [comment]
One thing we might consider is using the new standardized dependency injection interface (from https://github.com/aspnet/DependencyInjection). It is starting to be used by major frameworks, as well as ASP.NET Core and Entity Framework Core. The interface is basically:

```
  /// <summary>Defines a mechanism for retrieving a service object; that is, an object that provides custom support to other objects.</summary>
  /// <filterpriority>2</filterpriority>
  public interface IServiceProvider
  {
    /// <summary>Gets the service object of the specified type.</summary>
    /// <returns>A service object of type <paramref name="serviceType" />.-or- null if there is no service object of type <paramref name="serviceType" />.</returns>
    /// <param name="serviceType">An object that specifies the type of service object to get. </param>
    /// <filterpriority>2</filterpriority>
    object GetService(Type serviceType);
  }
```

What I think _should_ happen is that we should register the types that we want to inject into `ObjectGraphType` or any other graph type, and let the internals of this framework handle constructing the graph types with the types that the user provides.

**Comment by drobertson123** [comment]
I would like to add that it would be good to keep the internal dependencies of GraphQL as independent of the DI system as possible. We should be able to create a JSON version of the Schema and register the data sources using that. 

Ideally this would work by injecting schema fragments into GraphQL that let the service know what and how those registered objects would provide data. Each data source would do this independently and GraphQL would assemble the overall Schema for the service.

Queries and Mutations could be sent to the Data Source objects in the same way. GraphQL's primary responsibility would be parsing the requests, routing the requests to the proper sources and consolidating the data for the reply.

GraphQL Data Sources would need to comply with a set of Interfaces. These could be published in a completely independent GraphQL Data Source SDK project that would be used both in GraphQL and in any outside implementation of a Data Source object.

If you keep the SDK limited to just Interface definitions and some minor utility code you could drop the dependency requirements for the SDK to a very low level. This would decouple GraphQL from the Data Source objects and allow the developer a lot of leeway.

It would be good to avoid having lots of GraphQL internal objects required for the implementation of a Data Source. That would make it harder for both sides to change later. 

Sorry for the essay, just my 2 cents.

**Comment by joemcbride** [comment]
If I'm understanding it all, I'm agreeing with @drobertson123 .  We're moving towards an "instanced" based api instead of a type-based api (https://github.com/graphql-dotnet/graphql-dotnet/pull/91, https://github.com/graphql-dotnet/graphql-dotnet/pull/181, and I have a new PR in the works as well).  Moving towards an instanced-api will make it easier to build schemas on the fly and would alleviate the GraphQL lib from any type-building concerns.  In short, the core GraphQL engine will not be building any types.  Those would need to be created by the user of the library and passed to the GraphQL engine.

That being said, I would expect there to be some other library classes provided that could help building a schema from either C# types or perhaps the defined schema language, plugging in the resolvers.  Those other classes could have some DI hooks.

Examples of what I'm thinking:
* https://github.com/matthewmueller/graph.ql
* `makeExecutableSchema` from http://dev.apollodata.com/tools/graphql-tools/generate-schema.html

Here's a spike of a "Schema Creator" I made that can take in list of C# types.  Going the opposite route of using the text-schema and adding resolvers is not currently possible with the type-based system as we have it.

```csharp
public interface IPerson
{
    Guid Id { get; set; }
    string Name { get; set; }
}

public class Person : IPerson
{
    public Guid Id { get; set; }
    public string Name { get; set; }
}

public class Query
{
    public IPerson Person()
    {
        return new Person { Id = Guid.NewGuid(), Name = "Quinn"};
    }
}

var types = new[] { typeof(IPerson), typeof(Person), typeof(Query) };
var schema = TypedSchema.For(types);
var printer = new SchemaPrinter(schema);
Console.WriteLine(printer.Print());
var result = await _runner.ExecuteAsync(schema, @"
    query {
      person {
        id
        name
      }
    }
");
Console.WriteLine();
Console.WriteLine(result);
```

```
interface IPerson {
  id: ID!
  name: String!
}

type Person implements IPerson {
  id: ID!
  name: String!
}

type Query {
  person: IPerson
}

{
  "data": {
    "person": {
      "id": "e86547f6-d77a-48c3-8849-6f205b5a38cc",
      "name": "Quinn"
    }
  }
}
```

**Comment by cable729** [comment]
I like the idea of moving away from an inheritance and constructor-based setup. This looks good, but may I suggest a few improvements?

```
public class Person
{
    public Guid Id { get; set; }
    public string Name { get; set; }
}

public class MyQuery
{
    private readonly IPersonRepository _people;

    public MyQuery(IPersonRepository people) { _people = people; }

    public IPerson Person(int id)
    {
        return _people.Find(id);
    }
}

// using Microsoft.Extensions.DependencyInjection
IServiceProvider dependencyProvider = new ServicesCollection()
    .AddTransient<IPersonRepository, PersonRepository>()
    .BuildServiceProvider();

var schema = TypedSchema.For<MyQuery>(dependencyProvider);
// or: var schema = TypedSchema.For<MyQuery, MyMutation>(...);

var runner = Runner.For(schema);
const string getPersonQuery =@"
    query {
      person('e86547f6-d77a-48c3-8849-6f205b5a38cc') {
        id
        name
      }
    }";

var result = await _runner.ExecuteAsync(getPersonQuery);
Console.WriteLine(result);
// same as your result

const string queryWithParameters = "...";
parameters = "{ personId = 'e86547f6-d77a-48c3-8849-6f205b5a38cc }";
// or parameters = new Dictionary<string, string>() { ["personId"] = "e86547f6-d77a-48c3-8849-6f205b5a38cc" };
var result2 = await _runner.ExecuteAsync(queryWithParameters, parameters);

Assert.Equals(result, result2);
// true
```

Attributes would also make this very nice to use.

The only thing I'm not exactly sure on is where the `ExecutionContext` should go.. It could be added as a parameter to any method that uses it I suppose, and other methods that don't want it don't have to add it to the parameter list.

**Comment by joemcbride** [comment]
Made a few updates related to this in #362

**Comment by joemcbride** [comment]
Going to close this.  An interface was added in #362 and is available in the 2.0 alpha.
